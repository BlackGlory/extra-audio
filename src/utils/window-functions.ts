export type WindowFunction = (alpha: number) => number

export const rectangleWindow: WindowFunction = () => 1

export const triangularWindow: WindowFunction = alpha => {
  return 1 - Math.abs(2 * alpha - 1)
}

export const hannWindow: WindowFunction = alpha => {
  return 0.5 * (1 - Math.cos(2 * Math.PI * alpha))
}

export const hammingWindow: WindowFunction = alpha => {
  return 25 / 46 - (21 / 46) * Math.cos(2 * Math.PI * alpha)
}
