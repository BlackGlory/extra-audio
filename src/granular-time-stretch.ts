import { getSample } from './utils/get-sample.js'
import { triangularWindow, WindowFunction } from './utils/window-functions.js'

export function granularTimeStretch(
  input: Float32Array
, output: Float32Array
, grainSize: number = 480
, window: WindowFunction = triangularWindow
, overlapRatio: number = 0.5
): Float32Array {
  // outputIndex / inputIndex = outputLength / inputLength
  const scale = output.length / input.length

  // 这是假设输出粒子不重叠时的粒子数量.
  // 在`overlapRatio`参数大于0的时候, 输出的粒子会重叠, 实际产生的粒子数量会多于此数值.
  const outputGrains = output.length / grainSize

  const step = 1 - overlapRatio
  for (
    let outputGrainIndex = 0
  ; outputGrainIndex < outputGrains
  ; outputGrainIndex += step
  ) {
    const inputGrainIndex = outputGrainIndex / scale

    const inputIndexStart = inputGrainIndex * grainSize
    const outputIndexStart = Math.floor(outputGrainIndex * grainSize)

    for (let offset = 0; offset < grainSize; offset++) {
      const outputIndex = outputIndexStart + offset
      const inputIndex = inputIndexStart + offset
      output[outputIndex] += getSample(input, inputIndex)
                           * window(offset / (grainSize - 1))
    }
  }

  return output
}
