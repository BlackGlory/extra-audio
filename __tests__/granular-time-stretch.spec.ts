import { granularTimeStretch } from '@src/granular-time-stretch.js'
import { rectangleWindow } from '@utils/window-functions.js'

describe('granularTimeStretch', () => {
  it('output.length === input.length', () => {
    const input = new Float32Array([0, 1, 2, 3])
    const output = new Float32Array(input.length)

    const result = granularTimeStretch(input, output, 1, rectangleWindow, 0)

    expect(result).toBe(output)
    expect(result).not.toBe(input)
    expect(result).toEqual(input)
  })

  it('output.length > input.length', () => {
    const input = new Float32Array([1, 2])
    const output = new Float32Array(4)

    const result = granularTimeStretch(input, output, 1, rectangleWindow, 0)

    expect(result).toBe(output)
    expect(result).not.toBe(input)
    expect(result).toEqual(new Float32Array([1, 1.5, 2, 2]))
  })

  it('output.length < input.length', () => {
    const input = new Float32Array([1, 1, 2, 2])
    const output = new Float32Array(input.length / 2)

    const result = granularTimeStretch(input, output, 1, rectangleWindow, 0)

    expect(result).toBe(output)
    expect(result).not.toBe(input)
    expect(result).toEqual(new Float32Array([1, 2]))
  })
})
