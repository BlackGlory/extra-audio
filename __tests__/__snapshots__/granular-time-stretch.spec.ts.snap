// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`granularTimeStretch > output.length < input.length 1`] = `
Float32Array [
  1,
  2,
]
`;

exports[`granularTimeStretch > output.length > input.length 1`] = `
Float32Array [
  1,
  1.5,
  2,
  2,
]
`;
